Bulk Pubsub Optimization Opportunities (#3808)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/pubsub/pubsub-bulk.md

{{< /tabs >}}
## How components handle publishing and subscribing to bulk messages

> When talking about event publish/subscribe, there are two kind of network transfers involved.

Reviewer: ```suggestion
For event publish/subscribe, two kinds of network transfers are involved.
```

1. From/To *App* To/From *Dapr*.

> 2. From/To *Dapr* To/From *Pubsub Broker*.

Reviewer: ```suggestion
1. From/To *Dapr* To/From *Pubsub Broker*.
```

> And, these are the opportunities where optimization is possible for multiple Single calls v/s Bulk. Here, optimization means that Bulk calls are sent, which reduce number of overall calls and thus increased throughput ad better latency.

Reviewer 2: ```suggestion
These are the opportunities where optimization is possible. When optimized, a Bulk requests are, which reduce number of overall calls and thus increase throughput and provide better latency.
```

> And, these are the opportunities where optimization is possible for multiple Single calls v/s Bulk. Here, optimization means that Bulk calls are sent, which reduce number of overall calls and thus increased throughput ad better latency.

Reviewer: Adding to Mukundan's edit ^^

```suggestion
These are the opportunities where optimization is possible. When optimized, a Bulk request is sent, which reduces the number of overall calls, thus increasing throughput and providing better latency.
```

> On enabling Bulk Publish and/or Bulk Subscribe, Point 1 i.e. From/To *App* To/From *Dapr* is optimized for All components (depending on whether either one of Bulk Publish/ Bulk Subscribe OR Both are enabled).

Reviewer 2: ```suggestion
On enabling Bulk Publish and/or Bulk Subscribe, the communication between the App and Dapr sidecar (Point 1 above) is optimized for **all components**.
```

> But, optimization of Point 2 i.e. From/To *Dapr* To/From *Pubsub Broker* would depend on a number of factors, i.e. if Broker inherently supports Bulk Publish/Subscribe and if Dapr component is updated to support Batching. Currently, following components are updated to support this level of optimization:

Reviewer 2: ```suggestion
Optimization from Dapr sidecar to the pub/sub broker would depend on a number of factors, for example:
- If the broker inherently supports Bulk pub/sub
- If the Dapr component is updated to support the use of bulk APIs provided by the broker. 

Currently, the following components are updated to support this level of optimization:
```

<table width="100%">
    <tr>
        <th>Component</th>
        <th>Bulk Publish</th>
        <th>Bulk Subscribe</th>
    </tr>
    <tr>
        <td>Kafka</td>
        <td>Yes</td>
        <td>Yes</td>
    </tr>
    <tr>
        <td>Azure Servicebus</td>
        <td>Yes</td>
        <td>Yes</td>
    </tr>
    <tr>
        <td>Azure Eventhubs</td>
        <td>Yes</td>
        <td>Yes</td>
    </tr>

> </table>

Reviewer 2: should this table be added as part of the overall comp ref table? And then have a link pointing to the above doc? 
Additionally, adding a diagram denoting the points of optimization would be great.

Reviewer: nit - This table should probably be markdown instead of HTML

Reviewer 3: Hannah - Can you help draw a diagram above with me to explain these point. Much better than text.
